quinn-proto/src/connection/assembler.rs

```diff
@@ -197,6 +233,48 @@ impl Chunk {
 struct Buffer {
     offset: u64,
     bytes: Bytes,
+    allocation_size: usize,
+    defragmented: bool,
+}
+
+impl Buffer {
+    /// Constructs a new fragmented Buffer
+    fn new(offset: u64, bytes: Bytes, allocation_size: usize) -> Self {
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: false,
+        }
+    }
+
+    /// Constructs a new defragmented Buffer
+    fn new_defragmented(offset: u64, bytes: Bytes) -> Self {
+        let allocation_size = bytes.len();
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: true,
+        }
+    }
+
+    /// Discards data before `offset` and flags `self` as defragmented if it has good utilization
+    fn try_mark_defragment(&mut self, offset: u64) {
+        let duplicate = offset.saturating_sub(self.offset) as usize;
+        self.offset = self.offset.max(offset);
+        if duplicate >= self.bytes.len() {
+            self.bytes = Bytes::new();
+            self.defragmented = true;
+        } else {
+            self.bytes.advance(duplicate);
+        }
+        self.defragmented = self.defragmented || self.bytes.len() * 6 / 5 >= self.allocation_size;
+        if self.defragmented {
+            // Make sure that defragmented buffers do not contribute to over-allocation
```

Comment:
This logic seems more complicated than necessary. If we've established that `self.defragmented = true` because the chunk is now empty, we should just reset the allocation size and return immediately.

Reply:
Fixed.